src/main/java/org/spongepowered/common/world/storage/WorldStorageUtil.java

```diff
@@ -0,0 +1,114 @@
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.common.world.storage;
+
+import com.flowpowered.math.vector.Vector3i;
+import net.minecraft.nbt.CompressedStreamTools;
+import net.minecraft.nbt.NBTTagCompound;
+import net.minecraft.world.WorldServer;
+import net.minecraft.world.chunk.storage.IChunkLoader;
+import net.minecraft.world.chunk.storage.RegionFile;
+import net.minecraft.world.chunk.storage.RegionFileCache;
+import org.spongepowered.api.data.DataContainer;
+import org.spongepowered.common.data.persistence.NbtTranslator;
+import org.spongepowered.common.data.util.NbtDataUtil;
+import org.spongepowered.common.interfaces.world.IMixinAnvilChunkLoader;
+import org.spongepowered.common.scheduler.SpongeScheduler;
+
+import java.io.DataInputStream;
+import java.io.File;
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.Collections;
+import java.util.Optional;
+import java.util.concurrent.CompletableFuture;
+
+public class WorldStorageUtil {
+
+    public static CompletableFuture<Boolean> doesChunkExist(WorldServer world, IChunkLoader chunkLoader, Vector3i chunkCoords) {
+        int x = chunkCoords.getX();
+        int z = chunkCoords.getZ();
+        if (!(chunkLoader instanceof IMixinAnvilChunkLoader) || !SpongeChunkLayout.instance.isValidChunk(x, chunkCoords.getY(), z)) {
+            return CompletableFuture.completedFuture(false);
+        }
+        return SpongeScheduler.getInstance().submitAsyncTask(() -> ((IMixinAnvilChunkLoader) chunkLoader).chunkExists(world, x, z));
+    }
+
+    public static CompletableFuture<Optional<DataContainer>> getChunkData(WorldServer world, IChunkLoader chunkLoader, Vector3i chunkCoords) {
+        int x = chunkCoords.getX();
+        int y = chunkCoords.getY();
+        int z = chunkCoords.getZ();
+        if (!(chunkLoader instanceof IMixinAnvilChunkLoader) || !SpongeChunkLayout.instance.isValidChunk(x, y, z)) {
+            return CompletableFuture.completedFuture(Optional.empty());
+        }
+        File worldDir = ((IMixinAnvilChunkLoader) chunkLoader).getWorldDir().toFile();
+        return SpongeScheduler.getInstance().submitAsyncTask(() -> {
+            DataInputStream stream = RegionFileCache.getChunkInputStream(worldDir, x, z);
+            return Optional.ofNullable(readDataFromRegion(stream));
+        });
+    }
+
+    public static DataContainer readDataFromRegion(DataInputStream stream) throws IOException {
+        if (stream == null) {
+            return null;
+        }
+        NBTTagCompound data = CompressedStreamTools.read(stream);
+
+        if (!data.hasKey("Level", NbtDataUtil.TAG_COMPOUND)) {
```

Comment:
If this is a verbatim copy from `AnvilChunkLoader`, then I'd recommend following [this](https://github.com/SpongePowered/SpongeCommon/blob/1.9/src/main/java/org/spongepowered/common/mixin/core/entity/player/MixinEntityPlayer.java#L286-L461) as a guide.